src/sign.rs

```diff
@@ -0,0 +1,137 @@
+use anyhow::Error;
+use chrono::Utc;
+use pgp::{
+    armor::BlockType,
+    crypto::HashAlgorithm,
+    packet::{Packet, SignatureConfig, SignatureType, SignatureVersion, Subpacket},
+    types::{KeyTrait, SecretKeyTrait},
+    Deserializable, SignedSecretKey,
+};
+use rayon::prelude::*;
+use sha2::Digest;
+use std::{
+    fs::File,
+    path::{Path, PathBuf},
+    time::Instant,
+};
+
+use crate::config::Config;
+
+pub(crate) struct Signer {
+    gpg_key: SignedSecretKey,
+    gpg_password: String,
+}
+
+impl Signer {
+    pub(crate) fn new(config: &Config) -> Result<Self, Error> {
+        let mut key_file = File::open(&config.gpg_key_file)?;
+        let gpg_password = std::fs::read_to_string(&config.gpg_password_file)?;
+        Ok(Signer {
+            gpg_key: SignedSecretKey::from_armor_single(&mut key_file)?.0,
+            gpg_password,
+        })
+    }
+
+    pub(crate) fn sign_directory(&self, path: &Path) -> Result<(), Error> {
+        let mut paths = Vec::new();
+        for entry in std::fs::read_dir(path)? {
+            let entry = entry?;
+            let path = entry.path();
+
+            if !entry.metadata()?.is_file() || should_exclude_path(&path) {
+                continue;
+            }
+            paths.push(path);
+        }
+
+        self.sign_batch(&paths)
+    }
+
+    fn sign_batch(&self, paths: &[PathBuf]) -> Result<(), Error> {
+        let start = Instant::now();
+        println!(
+            "hashing and signing {} files across {} threads",
+            paths.len(),
+            rayon::current_num_threads().min(paths.len())
+        );
+
+        paths
+            .par_iter()
+            .map(|path| self.sign(path))
+            .collect::<Result<Vec<()>, Error>>()?;
+
+        println!(
+            "finished hashing and signing {} files in {:.2?}",
+            paths.len(),
+            start.elapsed()
+        );
+
+        Ok(())
+    }
+
+    fn sign(&self, path: &Path) -> Result<(), Error> {
+        let data = std::fs::read(path)?;
+
+        self.generate_sha256(path, &data)?;
+        self.gpg_sign(path, &data)?;
```

Comment:
I think this is pre-existing and we've discussed it, but maybe a comment here would be helpful for the future -- we want to be signing the sha256 files for speed, right? Maybe it doesn't matter in practice though

Reply:
Do you mean documenting the current behavior, or documenting the future behavior?

Reply:
Either; I mostly want a comment that makes it clear that we do indeed currently hash twice and sign one of those (since AFAICT the gpg signature is still a signature on a hash)

Reply:
So, we indeed hash twice, but the hash used by gpg is not a normal hash of just the contents we're signing: the hash also includes other stuff in it, like the creation date or the key issuing the signature. The `pgp` has an [example of signing the hash directly](https://docs.rs/pgp/0.7.1/pgp/composed/signed_key/index.html), but as that page says the signature it generates is *not* RFC-compliant. When I tried implementing it gpg rejected the resulting signature.

I guess if someone spends time to dig into the RFC a way to hash only one time, a solution could be found. I don't have that knowledge though, and studying the RFC for that is not worth the effort IMO. I'll add a comment.